File: src/leetCode/Ex42.java
package leetCode;

import java.util.Deque;
import java.util.LinkedList;

public class Ex42 {
    public int trap(int[] height) {
        Deque<Integer> stack = new LinkedList<>();
        int res = 0;
        for (int i = 0; i < height.length; i++) {
            while (!stack.isEmpty() && height[i] > height[stack.peek()]) {
                int top = stack.pop();
                if (stack.isEmpty()){
                    break;
                }
                int left = stack.peek();
                int width = i-left-1;
                res += width*(Math.min(height[left],height[i])-height[top]);

            }
            stack.push(i);
        }
        return res;
    }
}

File: src/leetCode/ex42.kt
package leetCode

fun trap(height: IntArray): Int {
    var left= 0
    var right = height.size-1
    var leftMax = 0
    var rightMax = 0
    var res = 0
    while (left<right){
        if (height[left]<height[right]){
            if (height[left]<leftMax){
                res+=leftMax-height[left]
            }else{
                leftMax=height[left]
            }
            left++
        }else{
            if (height[right]<rightMax){
                res+=rightMax-height[right]
            }else{
                rightMax=height[right]
            }
            right--
        }
    }
    return res
}

fun main(args: Array<String>) {
    trap(intArrayOf(0,1,0,2,1,0,1,3,2,1,2,1)).apply(::println)
}